Allow to select the instance and the main method to execute when launching the interpreter (#60, #86)

echebbi opened the pull request:
### Contribution

Fix #60 and fix #86 by:
 - improving the Launch Configuration tab to allow a user to select a `@main` method and the element of the model on which it should be called
 - asking the user to select a couple (method / element) when several `@main` methods / target elements are available and the user did not specified the ones to use
 - persisting the launch configuration created by the contextual menu (right-click > Run As) so that it can thereafter be run with CTRL+F11 (on Windows).


### Comments

I did not implemented new automatic tests, but I manually tested various scenarios and it seems to work as expected.

This PR is a draft because I'd like to improve the user documentation before merging.

### Screenshots 

#### The new Launch Configuration tab

<div align="center">
<img src="https://user-images.githubusercontent.com/25926433/70394304-49420700-19f4-11ea-8260-d7fcad16e5e5.png"/>
</div>

#### The main method selection dialog:

<div align="center">
<img src="https://user-images.githubusercontent.com/25926433/70394353-bb1a5080-19f4-11ea-9e75-e3dae23e5c6f.png"/>
</div>

#### The main model element selection dialog:

<div align="center">
<img src="https://user-images.githubusercontent.com/25926433/70394362-d2593e00-19f4-11ea-8f53-faca6c4b1f70.png"/>
</div>

#### The entry point selection dialog:

(called when ALE is executed but an ambiguity is detected)

<div align="center">
<img src="https://user-images.githubusercontent.com/25926433/70539126-87116d80-1b63-11ea-8782-6e41ca6dd92f.png"/>
</div>

> _Note:_ the choise is persisted so that the user won't have to make it again for this launch configuration

dvojtise commented on plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/launchconfig/UiUtils.java:
```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
```
This class looks useful, maybe consider having a copy in GEMOC commons too

echebbi replied:
In the [org.eclipse.gemoc.commons.eclipse.ui](https://github.com/eclipse/gemoc-studio/tree/f413e861a80678eb9d4f55268d032f9c31f90f6a/commons/plugins/org.eclipse.gemoc.commons.eclipse.ui) bundle?

dvojtise commented on plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/internal/DatedDslSemantics.java:
```diff
@@ -0,0 +1,96 @@
+ * An instance of this class is not updated as the underlying resource
+ * defining the semantics (e.g. a <i>.dsl</i> or a <i>.ale</i> file) changes.
+ */
+public class DatedDslSemantics implements DslSemantics {
```
Maybe the name of this class is misleading
Date implies to have a date information in the class but there is not
 while the purpose of the class seems to have an immutable DslSemantic ?

echebbi replied:
Renamed as `ImmutableDslSemantics`

echebbi commented:
I updated the "entry point selection dialog" so that it now displays the name of the EObject in addition to its fully qualified name (see first comment for the screenshot).

I also renamed `DatedDslSemantics` as `ImmutableDslSemantics`.

Currently a lot of tests do not pass; that's because the `ALEInterpreter` does not automatically uses semantics' first `@main` method and model's first element anymore. Indeed, the method to call and the model element on which it will be called must be given as arguments (see [ALEInterpreter#L241](https://github.com/gemoc/ale-lang/pull/89/files#diff-165f5f16a5e9f2a8aa366838c2f5730b)). 

Ideally I'd like to keep it that way because that's not the business of the interpreter to make such choices and that's why I gave this reponsibility to `AleLaunchConfigurationDelegate`. I'm going to think about a solution, maybe based on the Strategy pattern to allow different behaviors (e.g. use first main method by default, or cancel launch when ambiguous, or prompt a dialog to ask for a selection). Anyhow, this will be part of ALE's public API so it must be easy to use.

echebbi commented:
The [tests now pass](https://ci.inria.fr/gemoc/job/ale-lang/job/enhance-launch-configuration/), so this PR can be merged.

### Important: introduces breaking changes:

- The only method to execute ALE's interpreter is now `ALEInterpreter.eval(EObject, Method, List<Object, DslSemantics)`

Although worth mentioning this change should be easy take over; see changes made to `examples/logo.standalone/src/logo/standalone/Main.java` for an example.

### Note for later

`ALEInterpreter.eval` fails when called with a `caller` that has been loaded in another EMF context (see JavaDoc for details). I think that it shows a lack of cohesion and that we should come up with a better design. Maybe update interpreter's `queryEnvironment` at the beginning of the execution with relevant EPackages rather than relying on the interpreter to create the semantics and load the model.
